source/game/field/BoxColManager.cc

```diff
@@ -0,0 +1,620 @@
+        bool setAdditionalFlag, void *userData) {
+    BoxColFlag flag = BoxColFlag(eBoxColFlag::ITEM);
+
+    if (setAdditionalFlag) {
+        flag.setBit(eBoxColFlag::FLAG_100);
+    }
+
+    return insert(radius, maxSpeed, pos, flag, userData);
+}
+
+/// @addr{0x80786078}
+BoxColUnit *BoxColManager::insertObject(f32 radius, f32 maxSpeed, const EGG::Vector3f *pos,
+        bool setAdditionalFlag, void *userData) {
+    BoxColFlag flag = BoxColFlag(eBoxColFlag::OBJECT_2);
+
+    if (setAdditionalFlag) {
+        flag.setBit(eBoxColFlag::FLAG_100);
+    }
+
+    return insert(radius, maxSpeed, pos, flag, userData);
+}
+
+/// @addr{0x80786120}
+BoxColUnit *BoxColManager::insertDrivable(f32 radius, f32 maxSpeed, const EGG::Vector3f *pos,
+        bool setAdditionalFlag, void *userData) {
+    BoxColFlag flag = BoxColFlag(eBoxColFlag::DRIVABLE);
+
+    if (setAdditionalFlag) {
+        flag.setBit(eBoxColFlag::FLAG_100);
+    }
+
+    return insert(radius, maxSpeed, pos, flag, userData);
+}
+
+/// @addr{0x80786DBC}
+void BoxColManager::reinsertUnit(BoxColUnit *unit) {
+    f32 radius = unit->m_radius;
+    f32 maxSpeed = unit->m_range - radius;
+    const EGG::Vector3f *pos = unit->m_pos;
+    BoxColFlag flag = unit->m_flag;
+    void *userData = unit->m_userData;
+
+    remove(&unit);
+    insert(radius, maxSpeed, pos, BoxColFlag(), userData)->m_flag = flag;
+}
+
+/// @addr{0x80786774}
+void BoxColManager::search(BoxColUnit *unit, const BoxColFlag &flag) {
+    searchImpl(unit, flag);
+    resetIterators();
+}
+
+/// @addr{0x80786B14}
+void BoxColManager::search(f32 radius, const EGG::Vector3f &pos, const BoxColFlag &flag) {
+    searchImpl(radius, pos, flag);
+}
+
+/// @addr{0x807855DC}
+BoxColManager *BoxColManager::CreateInstance() {
+    assert(!s_instance);
+    s_instance = new BoxColManager;
+    return s_instance;
+}
+
+/// @addr{0x8078562C}
+void BoxColManager::DestroyInstance() {
+    assert(s_instance);
+    delete s_instance;
+    s_instance = nullptr;
+}
+
+BoxColManager *BoxColManager::Instance() {
+    return s_instance;
+}
+
+/// @brief Helper function since the getters share all code except the flag.
+void *BoxColManager::getNextImpl(s32 id, const BoxColFlag &flag) {
+    if (id == MAX_UNIT_COUNT) {
+        return nullptr;
+    }
+
+    BoxColUnit *unit = m_units[id];
+    iterate(id, flag);
+
+    return unit->m_userData;
+}
+
+/// @addr{Inlined}
+void BoxColManager::iterate(s32 &iter, const BoxColFlag &flag) {
+    while (++iter < m_maxID) {
+        if (m_units[iter]->m_flag == flag) {
+            return;
+        }
+    }
+
+    iter = MAX_UNIT_COUNT;
+}
+
+/// @addr{0x80786134}
+BoxColUnit *BoxColManager::insert(f32 radius, f32 maxSpeed, const EGG::Vector3f *pos,
+        const BoxColFlag &flag, void *userData) {
+    if (m_unitCount >= MAX_UNIT_COUNT) {
+        return nullptr;
+    }
+
+    s32 unitID = m_nextUnitID;
+    BoxColUnit *pUnit = &m_unitPool[unitID];
```

Comment:
```suggestion
    BoxColUnit &unit = m_unitPool[unitID];
```
Returns can safely use `&unit` to get the pointer to the unit.